packages/wast-parser/src/tokenizer.js

```diff
@@ -165,6 +175,144 @@ function tokenize(input: string) {
     char = input[current];
   }
 
+  // FSM helper functions
+
+  /**
+   * Creates a transition function mapping all characters matching a regular expression to a new state
+   *
+   * @param {RegExp} regex The regular expression to be matched against
+   * @param {String} nextState The state to transition to
+   * @param {Number} n How many characters to consume (default 1)
+   * @param {String} allowSeparators If the numeric separator "_" is allowed, this argument should contain the state to go to (default false)
+   *
+   * @return {Function} A state transition function which reads current state and input characters from its closure
+   */
+  const regexToState = (
+    regex,
+    nextState,
+    n = 1,
+    allowSeparators = false
+  ) => () => {
+    if (allowSeparators) {
+      if (char === NUMERIC_SEPARATOR) {
+        if (regex.test(lookbehind())) {
+          // Consume the separator and stay in current state
+          return [allowSeparators, 1];
+        } else {
+          unexpectedCharacter();
+        }
+      }
+    }
+
+    if (regex.test(lookahead(n, 0))) {
+      return [nextState, n];
+    }
+
+    return false;
+  };
+
+  /**
+   * Combines a list of transition functions into a complete state transition function.
+   * When none of the given transitions apply, the given default state is returned.
```

Comment:
Is it worth adding that the order of transitions matters? i.e. the first that matches is the 'winner'

Reply:
Definitely, that is important. Thanks Ill add it